File: app/src/main/java/com/admin/theapp/base/ViewModelFactory.java
package com.admin.theapp.base;

import android.arch.lifecycle.ViewModel;
import android.arch.lifecycle.ViewModelProvider;
import android.support.annotation.NonNull;
import android.support.v4.util.ArrayMap;

import com.admin.theapp.dagger.ViewModelSubComponent;
import com.admin.theapp.viewmodel.HotelDetailsViewModel;
import com.admin.theapp.viewmodel.HotelsViewModel;

import java.util.Map;
import java.util.concurrent.Callable;

import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton
public class ViewModelFactory implements ViewModelProvider.Factory {
    @NonNull
    private final ArrayMap<Class, Callable<? extends ViewModel>> creators;

    @Inject
    public ViewModelFactory(@NonNull ViewModelSubComponent viewModelSubComponent) {
        creators = new ArrayMap<>();
        creators.put(HotelsViewModel.class, viewModelSubComponent::hotelsViewModel);
        creators.put(HotelDetailsViewModel.class, viewModelSubComponent::hotelDetailsViewModel);
    }

    @Override
    @NonNull
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        Callable<? extends ViewModel> creator = creators.get(modelClass);
        if (creator == null) {
            for (Map.Entry<Class, Callable<? extends ViewModel>> entry : creators.entrySet()) {
                if (modelClass.isAssignableFrom(entry.getKey())) {
                    creator = entry.getValue();
                    break;
                }
            }
        }
        if (creator == null) {
            throw new IllegalArgumentException("Unknown model class " + modelClass);
        }
        try {
            return (T) creator.call();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }
}

File: app/src/main/java/com/admin/theapp/base/ViewModelFactory.kt
package com.admin.theapp.base

import android.arch.lifecycle.ViewModel
import android.arch.lifecycle.ViewModelProvider
import android.util.ArrayMap
import com.admin.theapp.dagger.ViewModelSubComponent
import com.admin.theapp.viewmodel.HotelDetailsViewModel
import com.admin.theapp.viewmodel.HotelsViewModel
import java.util.concurrent.Callable
import javax.inject.Inject

class ViewModelFactory @Inject constructor(viewModelSubComponent: ViewModelSubComponent) : ViewModelProvider.Factory {

    private val creators = ArrayMap<Class<out ViewModel>, Callable<out ViewModel>>()

    init {
        creators.put(HotelsViewModel::class.java, Callable { viewModelSubComponent.hotelsViewModel() })
        creators.put(HotelDetailsViewModel::class.java, Callable { viewModelSubComponent.hotelDetailsViewModel() })
    }

    override fun <T : ViewModel?> create(modelClass: Class<T>): T {
        var creator = creators[modelClass]
        if (creator == null) {
            creators.entries.forEach {
                if (modelClass.isAssignableFrom(it.key)) {
                    creator = it.value
                    return@forEach
                }
            }
        }
        try {
            return creator!!.call() as T
        } catch (e: Exception) {
            throw RuntimeException(e)
        }
    }
}
